Feature-auth-service

AmjadMassaoud opened the pull request:
@UladzislauKudravets This pull request introduces the initial setup and core functionality for the auth-service, a dedicated microservice for handling user authentication, authorization, and related operations within the Innogram platform.

The auth-service functions as an internal microservice responsible for handling authentication and authorization tasks. It is not directly exposed to users but is accessible through the API Gateway, which acts as a centralized entry point for incoming requests. The API Gateway serves as a client, routing user requests to various backend services, likethe auth-service, posts-service, chats-service. Ensuring secure access control, request management, and load balancing.

Description
The auth-service provides a robust set of features including:

User Authentication:
Standard email/password signup and login.
Secure logout mechanism.
JWT-based authentication with access and refresh tokens.
Access tokens are short-lived.
Refresh tokens are stored in httpOnly, secure cookies with SameSite=None.
Endpoint for validating access tokens.
Google OAuth 2.0 Integration:
Allows users to sign up or log in using their Google accounts.
Handles the Google OAuth callback and token exchange.
Password Management:
Secure password reset flow:
Request a password reset (generates a secure, time-limited token).
Reset password using the token sent to the user's email (email sending mechanism is outside this service's direct scope but facilitated by it).
Internal API Security:
Includes middleware to verify requests from other internal services using a shared secret (x-internal-api-secret header).
Database & Data Management:
Uses MongoDB as the primary data store, managed via TypeORM.
Defines TokenEntity for storing user session/token information (including refresh tokens and basic user details for authentication purposes) and PasswordResetTokenEntity for password reset operations.
Configuration:
Centralized configuration management using environment variables (.env file).
Joi validation for environment variables to ensure required settings are present and correctly formatted.
Configurable CORS policy.
API Structure:
RESTful API endpoints organized logically under /innogram/auth and /innogram/password.
Utilizes Express.js framework.
Business logic is encapsulated in provider modules.
Key Files & Modules
app.ts: Main Express application setup, middleware configuration, and route mounting.
main.ts: Server startup logic and database initialization.
configs/: Contains configurations for the server, CORS, cookies, TypeORM, and environment variable loading/validation.
controllers/: Defines API route handlers for authentication and password management.
providers/: (Assumed location for business logic, e.g., auth.provider.ts, password.provider.ts - these files were not provided but are referenced).
entities/: TypeORM entity definitions for database models.
middlewares/: Custom middleware, such as verify-internal-request.ts.
.env: Environment variable definitions.

AmjadMassaoud committed:
- fixed files structure

AmjadMassaoud committed:
- removed unnecessary files

AmjadMassaoud committed:
implement auth service with user registration, login, password reset, and token management

AmjadMassaoud committed:
feat(auth-service): update cookie-parser to version 1.4.7 and refactor authentication logic

- Added cookie-parser version 1.4.7 with updated LICENSE, README, and HISTORY files.
- Refactored authentication provider to use updated cookie settings for refresh tokens.
- Implemented password reset functionality with token generation and validation.
- Removed obsolete routes and types related to password management and email verification.
- Updated token verification logic to align with new user model structure.

AmjadMassaoud committed:
feat: integrate Google authentication with OAuth2 flow

- Added Google authentication provider to handle OAuth2 callback and token exchange.
- Implemented user creation and token generation for new Google users.
- Updated package.json to include google-auth-library dependency.
- Added utils for handling URL operations in the whatwg-url package.
- Updated package-lock.json to reflect new dependencies and versions.
- Enhanced access token verification process with logging for debugging.

AmjadMassaoud committed:
docs: add README for auth service with features, setup, and usage examples

AmjadMassaoud committed:
feat(auth-service): add PasswordResetTokenEntity and update password reset logic

AmjadMassaoud committed:
docs: update README with enhanced features, API endpoints, and usage examples

AmjadMassaoud committed:
feat(auth-service): enhance password reset token logic to return existing token if already issued

AmjadMassaoud committed:
chore(auth-service): update dependencies to include type definitions for express and redis

AmjadMassaoud committed:
feat(auth-service): implement Redis-based password reset token management and limit reset attempts

AmjadMassaoud committed:
Update README.md

AmjadMassaoud committed:
Refactor: Removed unwanted code

AmjadMassaoud committed:
Merge branch 'feature-auth-service' of https://github.com/AmjadMassaoud/Innogram into feature-auth-service

AmjadMassaoud committed:
Update .gitignore

Added node_module file to .gitignore file

AmjadMassaoud committed:
Delete apps/auth-service/node_modules directory

Deleted node_modules from auth-service

AmjadMassaoud committed:
Delete .vscode directory

- Deleted .vscode file

AmjadMassaoud committed:
Delete apps/chats-service directory

AmjadMassaoud committed:
Delete apps/notifications-service directory

AmjadMassaoud committed:
Delete apps/posts-service directory

AmjadMassaoud committed:
Delete apps/users-service directory

UladzislauKudravets commented on package.json:
```diff
@@ -30,19 +30,32 @@
     "test:e2e": "cross-env NODE_ENV=test jest --config ./test/jest-e2e.json"
   },
   "dependencies": {
```
do we really need these dependencies here at the top level directory? Consider cleaning up and moving deps to the microservice directory where they are really needed

UladzislauKudravets commented on apps/auth-service/app.ts:
```diff
@@ -0,0 +1,46 @@
+import express from 'express';
```
please don't forget to create internal git repo of auth-service according to the monorepo pattern

UladzislauKudravets replied:
comment related to the commit messages - please take a look at https://www.conventionalcommits.org/en/v1.0.0/. let's use it:)

UladzislauKudravets commented on apps/auth-service/README.md:
````diff
@@ -0,0 +1,198 @@
+POST /innogram/auth/validate-accessToken   # Validate access token
+```
+
+### Google OAuth Routes
+
+```http
+GET /innogram/auth/google-callback    # Google OAuth callback handler
+```
+
+### Password Management Routes
+
+```http
+POST /innogram/password/request-reset  # Request password reset
+POST /innogram/password/reset         # Reset password with token
+```
+
````
consider using openapi for describing your API

UladzislauKudravets commented on apps/auth-service/configs/cors.config.ts:
```diff
@@ -0,0 +1,55 @@
+) => {
+  if (err.message === 'Not allowed by CORS') {
+    res.status(httpStatus.FORBIDDEN).json({
+      error: 'Origin not allowed',
```
it's okay here but in the future please consider creating custom errors to avoid magic strings

UladzislauKudravets commented on apps/auth-service/controllers/auth.controller.ts:
```diff
@@ -0,0 +1,35 @@
+  handleRefreshToken(req, res).catch(next);
+});
+
+authController.post('/validate-accessToken', (req, res, next) => {
```
it's important to be consistent:) mixing of different naming pattern could be messy, please use /validate-access-token

UladzislauKudravets commented on apps/auth-service/entities/token-entity.ts:
```diff
@@ -0,0 +1,44 @@
+  googleUserId!: string;
+
+  @Column({ type: 'string' })
+  registrationMethod!: 'innogram' | 'google';
```
consider using an enum of possible reg methods

UladzislauKudravets commented on apps/auth-service/entities/token-entity.ts:
```diff
@@ -0,0 +1,44 @@
+} from 'typeorm';
+
+@Entity('tokens')
+export class TokenEntity {
```
seems that we have problem with naming here. Should the token entity include username, email etc?:)

UladzislauKudravets commented on apps/auth-service/entities/password-reset-token.entity.ts:
```diff
@@ -0,0 +1,26 @@
+} from 'typeorm';
+
+@Entity('password_reset_tokens')
+export class PasswordResetTokenEntity {
```
Consider renaming as well

UladzislauKudravets commented on apps/auth-service/entities/token-entity.ts:
```diff
@@ -0,0 +1,44 @@
+  id!: ObjectId;
+
+  @Column({ type: 'string' })
+  username!: string;
```
prefix 'user' seems redundant here

UladzislauKudravets commented on apps/auth-service/middlewares/verify-internal-request.ts:
```diff
@@ -0,0 +1,19 @@
+  ) {
+    return res
+      .status(403)
+      .json({ error: 'Unauthorized: Invalid internal API secret' });
```
custom error should be used

UladzislauKudravets commented on apps/auth-service/providers/auth.provider.ts:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    // Hash password
+    const hashedPassword = await bcrypt.hash(password, 10);
```
please consider using native 'crypto' library instead of bcrypt

UladzislauKudravets commented on apps/auth-service/providers/auth.provider.ts:
```diff
@@ -0,0 +1,272 @@
+  } catch (error) {
+    res
+      .status(httpStatus.INTERNAL_SERVER_ERROR)
+      .json({ error: 'Internal server error!!' });
```
custom errors

UladzislauKudravets commented on apps/auth-service/providers/auth.provider.ts:
```diff
@@ -0,0 +1,272 @@
+
+const TokenRepository = dataSource.getRepository(TokenEntity);
+
+export async function handleSignUp(req: Request, res: Response): Promise<void> {
```
req and res are abstructions of transport layer and should be used only in controllers. Files of the current layer should be named like services - they should be consisted of business logic only. All the interaction with db should be in repository layer

UladzislauKudravets commented on apps/auth-service/tsconfig.json:
```diff
@@ -0,0 +1,113 @@
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig to read more about this file */
+
+    /* Projects */
```
please remove commented code

AmjadMassaoud committed:
Update auth.controller.ts

Change route name for consistency

AmjadMassaoud committed:
Update password-reset-token.entity.ts

AmjadMassaoud committed:
Update token-entity.ts

AmjadMassaoud committed:
Rename token-entity.ts to user-auth-entity.ts

AmjadMassaoud committed:
Update auth.provider.ts

AmjadMassaoud committed:
Update tsconfig.json

removed unnecessary comments

AmjadMassaoud committed:
Create registration-method.ts

AmjadMassaoud committed:
Update user-auth-entity.ts

AmjadMassaoud committed:
feat(auth-service): refactor authentication logic and enhance error handling

AmjadMassaoud committed:
Merge branch 'feature-auth-service' of https://github.com/AmjadMassaoud/Innogram into feature-auth-service

AmjadMassaoud committed:
feat: Implement OpenAPI documentation for auth service

refactor: Refactor auth-service, implement necessary changes

Revised `README.md` to accurately reflect current service features, API endpoint definitions, project structure, and setup instructions.
